### DISCUSSION TITLE
MOOSE setting all my material properties to zero?

### DISCUSSION THREAD

**Discussion by richmondodufisan** [idaholab/moose#18409] [Q&A General] [answered]
In my material file, I declare the property using:

_len_scale(declareADProperty<Real>("length_scale")),

and after computing it in computeQpProperties, I added a print statement there- its outputting exactly what it's supposed to. Then in the kernel, I "get" this value with:

_len_scale(getADMaterialPropertyByName<Real>("length_scale")),

printing this outputs 0.000000. The same for all other material properties. I find this strange because its worked in the past. If I understand correctly, the string in quotes declared in the material file is what is being retrieved when I use "getADMaterialPropertyByName", right? The only different thing I could think of is that this is the first time I'm running a transient solve. 

The material property is a constant, btw. That is, what I computed in computeQpProperties depends only on values given from the user in the input file, so it's not supposed to change or be affected by anything else. 

I checked other materials, and they are all 0.000000 as well. I only caught this after I got an arithmetic exception somewhere else because of a division operation.

**Comment by GiudGiud** [comment, marked as answer]
Hello

Is _len_scale declared like this in the header?
const ADMaterialProperty<Real> & _len_scale ;
If len_scale is not declared as a reference, that would be the problem. But there's a guard (non-reference routine is private in MaterialProperty.h) against assigning it to a value so I doubt it's that. 

Guillaume

**Reply by richmondodufisan** [reply]
Hi,

Yes it's declared like this:

const ADMaterialProperty<Real> & _len_scale;

with a (Real) after ADMaterialProperty in angled brackets

**Reply by GiudGiud** [reply]
for both the material and the kernel right

**Reply by richmondodufisan** [reply]
Yes for both

**Reply by richmondodufisan** [reply]
But in the material, it's not const though, since I have to assign _len_scale[_qp] in computeQpProperties

**Reply by GiudGiud** [reply]
@rwcarlsen on mat props.
not sure what is going on here

**Reply by rwcarlsen** [reply]
Could you show more snippets of the code where you declare the property in the material, compute/assign the property,  retrieve the property in another object, and use the property in another object?  Are you using any non-default execute_on parameters for any objects, etc.?  Just need more context to help you diagnose.

**Reply by richmondodufisan** [reply]
_len_scale[_qp] = some calc involving _user_len_scale and some other params gotten with "getParam" from the input file
  
  printf("from user %f \n ",_user_len_scale);
  printf("from material %f  \n ",_len_scale[_qp]);

/// These print statements print exactly what I expect

///Kernel

/// header file declaration
const ADMaterialProperty<Real> & _len_scale;

/// source file constructor
_len_scale(getADMaterialPropertyByName<Real>("length_scale")),

printf("from kernel lenscale %f, init_porosity %f \n",_len_scale[_qp], _porosity_0[_qp]);

/// prints 0.00000 for both values

I also tried changing the "length_scale" in the getADMaterialPropertyByName to just a random string and I got an error, so it's apparently picking it from my previous declaration, I don't know why it's setting it to zero.

I also noticed something else this morning- a variable in my kernel, "porosity" is set to a constantIC, but when I printed it at the start of the computeQpResidual, it is also 0.00000 as well. But, in the .e file that's created, the initial value I set in the input file is shown correctly on the mesh.

I tried running it on a remote machine as well with far more computing power, but I got the same error.

**Reply by rwcarlsen** [reply]
Is the property declared as an AD property to begin with?  AD and non-AD properties are tracked separately.  So you need to declare+retrieve them consistently as AD and non-AD.  Also - just to confirm - the declaration for the member var is `const ADMaterialProperty<Real> & _len_scale;` - right?  Also just want to confirm - are your printfs in your kernel for the property values inside the computeQpResidual function?  Because inside the constructor or other functions, the properties will not be computed/ready.

**Reply by GiudGiud** [reply]
And _len_scale[_qp] = some calc is being done in computeValue of the material right?

**Reply by richmondodufisan** [reply]
Yes, the values from the user are initialized with ADReal, and then `getParam<Real>` to retrieve them. They're then calculated in computeQpProperties() (@GiudGiud, not computeValue(), is that the function I should be overriding?), and assigned to variables declared like `const ADMaterialProperty<Real> & _len_scale;` . Then, in each of the kernels, they are retrieved with `_len_scale(getADMaterialPropertyByName("length_scale")),`


And yes I added the print statement inside computeQpResidual(). I also added print statements to the other kernels, but the solver isn't getting to them yet.

![Capture](https://user-images.githubusercontent.com/52802063/126729003-949b2251-d644-4e02-8739-4278c22ed1af.JPG)

**Reply by richmondodufisan** [reply]
I just noticed- in the screenshot I sent above, everything including the user&material is zero. Yet, from earlier in the day, I had this: (The only thing I've done differently since then is to use a different mesh and changed the BCs and ICs. I also set _len_scale[_qp] = _user_len_scale directly to see if it was my calculation causing it (ignore the "etaqp", that was from checking another constant).

![Capture2](https://user-images.githubusercontent.com/52802063/126735461-9d480ed2-8997-4312-aa67-72dfd245e716.JPG)

If it's not consistent, do you think it has to do with my computer's processing power?

**Reply by GiudGiud** [reply]
no it has nothing to do with your computer processing power. 

Does the moose test suite pass on your machine?

computeQpProperties is the right method to override, computeValues doesn't exist my bad. 

I'm pretty clueless at this point. Could you please just send us or paste the source files?

**Reply by rwcarlsen** [reply]
Yes - there is probably a simple mistake lurking in either your material or kernel - and seeing the full code files would make this a lot easier to suss out.

**Reply by richmondodufisan** [reply]
Hi, I just discussed with my PI and there are some issues with my theoretical formulation I need to rectify, but can I send you the files privately so I can figure out what's wrong with the material system at least?

**Reply by GiudGiud** [reply]
Sure you can use the email on my github profile

**Reply by GiudGiud** [reply]
I get these compilation warnings:
```
/home/guillaume/projects/moose/framework/src/materials/ErosionDamageMaterial.C:42:54: warning: a temporary bound to 'ErosionDamageMaterial::_user_p_f' only persists until the constructor exits [-Wextra]
/home/guillaume/projects/moose/framework/src/materials/ErosionDamageMaterial.C:42:54: warning: a temporary bound to 'ErosionDamageMaterial::_user_g' only persists until the constructor exits [-Wextra]
/home/guillaume/projects/moose/framework/src/materials/ErosionDamageMaterial.C:42:54: warning: a temporary bound to 'ErosionDamageMaterial::_user_c' only persists until the constructor exits [-Wextra]
/home/guillaume/projects/moose/framework/src/materials/ErosionDamageMaterial.C:42:54: warning: a temporary bound to 'ErosionDamageMaterial::_user_eta_f' only persists until the constructor exits [-Wextra]
/home/guillaume/projects/moose/framework/src/materials/ErosionDamageMaterial.C:42:54: warning: a temporary bound to 'ErosionDamageMaterial::_user_len_scale' only persists until the constructor exits [-Wextra]
/home/guillaume/projects/moose/framework/src/materials/ErosionDamageMaterial.C:42:54: warning: a temporary bound to 'ErosionDamageMaterial::_user_porosity_0' only persists until the constructor exits [-Wextra]
/home/guillaume/projects/moose/framework/src/materials/ErosionDamageMaterial.C: In member function 'virtual void ErosionDamageMaterial::computeQpProperties()':
```

You need to change the constant declarations to   const ADReal _user_p_s; to store the value. 

Also this is more of an ADMaterial than a Material

And you have a 
if (i = 1) which will assign 1 to i

**Reply by richmondodufisan** [reply]
So in the material header file, the values gotten from the users should be initialized without references, but the declared materials (to be retrieved in the kernel) should be references?

then are you saying I should have DerivativeMaterialInterface<ADMaterial> ? I get a compiler error with this, what header do I need to add? The ADMaterial documentation (https://mooseframework.inl.gov/source/kernels/ADMaterial.html) isn't up yet.

I've corrected the first and third ones and am still getting the same thing as before when running my input file...

**Reply by GiudGiud** [reply]
yes
I m not sure why you are using DerivateMaterialInterface actually?
I don't see any material derivatives. 

I replaced <Material> with <ADMaterial> in the class template.

I was getting non-zero output on my side with those 3 changes

**Reply by richmondodufisan** [reply]
I thought I needed DerivativeMaterialInterface to use the AD System. Just to recap, you:

-declared the _user_len_scale and other variables gotten from the input file without references

-changed the classes I inherited from:
 in material file, from `DerivativeMaterialInterface<Material>` to just `Material`
in the kernel, from `DerivativeMaterialInterface<ADKernel>` to just `ADKernel`

-changed the assignment operator in the if statement

and you no longer got zero output? because I did those three and I'm still getting zeros

**Reply by GiudGiud** [reply]
No the AD system is totally separate from DerivativeMaterialInterface.
See the documentation, https://mooseframework.inl.gov/source/materials/DerivativeMaterialInterface.html, it's meant to help you use derivative of material properties, not directly compute the Jacobian from the residual

I did not remove the DerivativeMaterialInterface from your materials, just **specialized it to an ADMaterial**. However if you are not using DerivativeMaterialInterface, you might as well not use it. 

The last point is not important for this. 

I got non zero output from both the material and the kernel.

If you dont after those changes, maybe you are on an old version of moose? When did you last update?

**Reply by richmondodufisan** [reply]
I updated moose and specialized to ADMaterial  and I'm still getting zeros, only now its printing many more times than before. What arguments did you add to show those compiler warnings? The only ones I'm getting are -Wreorder warnings for not initializing in my src file in the same order as my header file.

![img](https://user-images.githubusercontent.com/52802063/126912535-7b50f2b7-b73f-4ac2-bff4-ba5ff9626e48.JPG)

**Reply by GiudGiud** [reply]
do they appear once you clean those up? I ll send you the files by email.
I made the rank4tensor a rank2tensor for testing as well

**Reply by richmondodufisan** [reply]
I just tried to compile these and I got errors- the message:

```
error: 'ADMaterial' was not declared in this scope; did you mean 'Material'?
   12 | class ErosionDamageMaterial : public DerivativeMaterialInterface<ADMaterial>
```

when I got this error message , I added the header `#include "ADMaterial.h"` before it compiled. Did this compile for you without doing that?

Then, I have 

```
*** ERROR ***
A 'VarCouplingMaterial' is not a registered object.

If you are trying to find this object in a dynamically linked library, make sure that
the library can be found either in your "Problem/library_path" parameter or in the
MOOSE_LIBRARY_PATH environment variable.
```

using the test input file you sent. Is this VarCouplingMaterial from MOOSE or is it a custom object? I updated MOOSE today according to https://mooseframework.inl.gov/getting_started/installation/update_moose.html, I'm on the latest version of Ubuntu (wsl).

**Reply by GiudGiud** [reply]
The input file is a test, you need to pass the --allow-test-object flag
yes it compiled for me.

**Reply by richmondodufisan** [reply]
you mean like this? `../farm-opt -i var_coupling.i --allow-test-object`

 because I'm still getting the same error. Also, any idea why mine isn't compiling without adding that header (that you didn't add)?

**Reply by GiudGiud** [reply]
--allow-test-object**s**

**Reply by GiudGiud** [reply]
About compiling, it's probably because I put all the files in the same folder, did not try to distinguish between kernel and materials, so the unity build (building a lot of the source files together) probably saved my compilation there.

**Reply by richmondodufisan** [reply]
same error

```
../farm-opt -i var_coupling.i --allow-test-objects



*** ERROR ***
A 'VarCouplingMaterial' is not a registered object.

If you are trying to find this object in a dynamically linked library, make sure that
the library can be found either in your "Problem/library_path" parameter or in the
MOOSE_LIBRARY_PATH environment variable.
```

**Reply by GiudGiud** [reply]
umm I thought it would have been included. Maybe it s because you're using your app not moose to run this input file. 
You can use moose with your materials 
OR
modify the test to get rid of the moose test objects
OR
move the test objects needed to your app

**Reply by richmondodufisan** [reply]
So I copied the test objects to my app, ran the test app and no longer got zero output.  Then I ran my input file, and also no longer got zero output. The only difference between our kernel files was that mine was registered to my app, and the modified one you sent was registered to moose. And a difference in printing output to the console but that shouldn't be relevant.

This was after I ended up deleting and recreating my app with the now updated version of moose. So apparently, apart from this (https://mooseframework.inl.gov/getting_started/installation/update_moose.html), I need to delete and recreate my app, and then recopy my custom kernels. Glad it's working now, thank you for your help!!

**Comment by rwcarlsen** [comment]
Could you show the code in your header and .C files where you define your material property member variable and retrieve it?  Sometimes people have silly mistakes like retrieving the material property and storing it into a material property _value_ instead of a _reference_ member variable.

[edit] - never mind - looks like you declared the member var right. Hmmm...
